force-app/main/default/classes/DMLOperations.cls

```diff
@@ -145,76 +179,150 @@ public with sharing class DMLOperations {
+        Account acc1 = new Account();
+        
+        // if the Account exists, update the description
+        for (Account acc : accList) {  
+            acc1 = acc;    
+            acc1.Description = 'Updated Account';
+        }
+        
+        // otherwise, create a new Account and set the description
+        if(accList.isEmpty()) {          
+            acc1.Name = accountName;
+            acc1.Description = 'New Account';
+        }
+        
+        upsert acc1;    
+        return acc1;
+    } 
 
     /**
      * Question 10
-	 * Manage the association between Contacts and Accounts
-	 * Ensure that for each Contact in the provided list, there is a corresponding Account Name that matches the Contact's last name.
-	 * If such an Account does not exist, the method creates it; if it does, the method ensures the Contact is associated with it. (Hint: use previous upsertAccount method - this may cause bad practices but for learning purposes it's okay.)
+     * Manage the association between Contacts and Accounts
+     * Ensure that for each Contact in the provided list, there is a corresponding Account Name that matches the Contact's last name.
+     * If such an Account does not exist, the method creates it; if it does, the method ensures the Contact is associated with it.
      * Example Contact.LastName = 'Doe' Link to Account.Name = 'Doe' 
      * Example Contact.LastName = 'Jane' Link to Account.Name = 'Jane'
-	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
-	 */
-	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
+     * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
+     */
+    public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
+        List<String> contNames = new List<String>();
+        String accName;
 		// Loop through each Contact
-
-		// Extract the Account Name from Contact's LastName
-
-		// Get the account based on the name or create one if it doesn't exist
-
-		// Link the Contact to its Account using the Account Id
-
-		// Upsert the Contacts
-	}
+        for ( Contact cont : contactsToUpsert) {
+            // Extract the Account Name from Contact's LastName
+            accName = cont.Lastname;
+            contNames.add(accName);
+        }
+        // Get the account based on the name or create one if it doesn't exist
+        List <Account> nameList = [SELECT Name, Id FROM Account WHERE Name = :contNames];
+
+        Map<String, Id> actMap = new Map<String, Id>();
+        for (Account acc : nameList) {
+            actMap.put(acc.Name, acc.Id);
+            // Link the Contact to its Account using the Account Id
+        } 
+        
+        for (Contact cont1 : contactsToUpsert){
+            cont1.AccountId = actMap.get(cont1.LastName);
+            if ( cont1.AccountId == null ){
+                Account newAcc = upsertAccount(cont1.LastName);
```

Comment:
Looks like you have a double for loop here, try to remove one. Remember upsertAccount already has a for loop on line 254